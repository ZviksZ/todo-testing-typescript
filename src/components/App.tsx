import * as React                                from 'react';
import {useReducer}                              from 'react';
import {Action, ActionType, ContextState, State} from "../types/stateType";
import NewTask                                   from "./NewTask";
import TasksList                                 from "./TasksList";
import {todoReducer}                             from "../reducers/todoReducer";

// Начальные значения стейта
export const initialState: State = {
   newTask: '',
   tasks: []
}

// <Partial> позволяет создать контекст без дефолтных значений
export const ContextApp = React.createContext<Partial<ContextState>>({});



const App: React.FC = () => {
// Используем созданный todoReducer, передав его аргументом в useReduser. Изначально в стейт попадёт initialState, и далее при диспатче (changeState) будет обновляться.
   const [state, changeState] = useReducer<React.Reducer<State, Action>>(todoReducer, initialState);

   const ContextState: ContextState = {
      state,
      changeState
   };

// Передаём в контекст результаты работы useReducer - стейт и метод его изменения
   return (
      <>
         <ContextApp.Provider value={ContextState}>
            <NewTask/>
            <TasksList/>
         </ContextApp.Provider>
      </>
   )
}

export default App;
