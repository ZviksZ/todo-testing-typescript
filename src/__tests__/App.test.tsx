import * as React from 'react';
import {shallow} from 'enzyme';
import {fireEvent, render, cleanup} from "@testing-library/react";
import App from "../components/App";

describe('<App />', () => {
// jest-функция afterEach с переданным коллбеком cleanup вызывается после каждого теста и очищает среду тестирования
   afterEach(cleanup);

   it('hasn`t got changes', () => {
//  метод shallow библиотеки enzyme позволяет производить юнит-тестирование, без отрисовки дочерних компонентов.
      const component = shallow(<App />);
// При первом запуске теста будет создан снимок компонента. При последующих тестированиях будет проверяться идентичность снимка с текущим содержимым компонента. Для обновления snapshots необходимо запустить тест с флагом -u: jest -u
      expect(component).toMatchSnapshot();
   });

// Так как в компоненте будут происходить асинхронные действия (вызываться события на DOM-элементах), оборачиваем тест в async
   it('should render right input value',  async () => {
// render() функция доступна в библиотеке @testing-library/react" и отличается от shallow() тем, что строит настоящее DOM-дерево для тестируемого компонента. Переменная container  — это элемент div, в который будет выведен компонент.
      const { container } = render(<App/>);
      expect(container.querySelector('input').getAttribute('value')).toEqual('');
// вызываем событие изменения инпута и передаём туда значение 'test'
      fireEvent.change(container.querySelector('input'), {
         target: {
            value: 'test'
         },
      })
// ожидаем получить в инпуте значение 'test'
      expect(container.querySelector('input').getAttribute('value')).toEqual('test');
// вызываем событие клика на кнопку. При этом событии поле инпута должно очищаться
      fireEvent.click(container.querySelector('button'))
// ожидаем получить в инпуте пустое значение атрибута value
      expect(container.querySelector('input').getAttribute('value')).toEqual('');
   });
})
