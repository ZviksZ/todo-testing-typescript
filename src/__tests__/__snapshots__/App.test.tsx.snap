// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> hasn\`t got changes 1`] = `
<Fragment>
  <ContextProvider
    value={
      Object {
        "changeState": [Function],
        "state": Object {
          "newTask": "",
          "tasks": Array [],
        },
      }
    }
  >
    <NewTask />
    <TasksList />
  </ContextProvider>
</Fragment>
`;
